README
Range::Object version 0.9
=========================

This suite of modules implements unified API for manipulating ranges of
objects of different kinds.

In terms of Range::Object namespace, a range is defined as a set of items,
either disjointed (individual) or contiguous, or a combination of separate
items and ranges. Intersecting or adjacent ranges are not supported directly
and will be collapsed silently into wider contiguous range.

Although Range::Object descendant module can store any number of separate
values (objects) and ranges, it is optimized for storing contiguous ranges
of arbitrary length with minimal memory and storage footprint; the other
effect of this being fast serialization and deserialization of Range::Object
instances. It cannot come without cost though; Range::Object uses more
CPU cycles than similar hash-based modules.

Good application for this kind of object storage can be an implementation
of user permission tables for large number of objects, especially if such
permissions are typically assigned in large contiguous ranges. For
example, if an user has read permission for objects 1-10000 and write
permission for objects 1-100, 200-300 and 1000-9999, storing these
identificators as hash keys is memory expensive, and can become
prohibive when number of tables and users increase. Compared to that
approach, Range::Object can be a reasonable compromise between memory
and CPU utilization.

This module also comes useful for user input validation when dealing with
HTML forms or something similar.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Test::More
  List::Util

COPYRIGHT AND LICENSE

Copyright (C) 2011 by Alexander Tokarev.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.6.0 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl module Range::Object.

0.92  Mon Nov 21 00:26:33 2011

    Removed versions from all packages except Range::Object itself.
    It appears that having all packages versioned only leads to
    confusion (mine, specifically).

0.91  Sun Nov 20 17:11:38 2011

    Apparently, permission for Range::String namespace is already
    taken on CPAN; to avoid collision I renamed my module to
    Range::Strings.

    No other changes were made.

0.9   Sun Nov 20 16:39:23 2011

    Cleaned up code and POD sections for CPAN release.

0.3   Wed Sep 14 01:04:47 2011

    Testing on 32-bit Perl build uncovered a few bugs, mostly with
    regards to digit handling. To avoid messing with integer logic,
    a new Range::DigitString module was added. Basically it's the
    same as Range::Extension except that it doesn't allow [*#]
    in the beginning.

    Fixed a bug in Range::Object that prevented an integer or
    digit string with zero value from being handled properly.

    Update test suite to handle Range::DigitString too.

0.25  Mon Sep 05 23:17:46 2011

    Fixed a bug in Range::Object: both add() and remove() returned
    failure if passed empty input range, when they should have
    returned $self. They do now.

0.24  Thu Aug 18 23:36:04 2011

    Fixed a bug in Range::PhoneDigits: ranges of numbers
    beginning with [*#] did not expand properly.
    Added a test for this condition as well.

    Abstracted expansion code as subroutine in Range::Object.

0.23  Mon Aug 15 17:10:04 2011

    Wrote extended set of tests involving valid and invalid input
    cases for each individual field type.

    Fixed all bugs uncovered by these tests. A lot of them.

0.2   Thu Aug 11 19:06:57 2011

    Fixed a wart in test suite: not all invalid input options were
    checked for each module, leading to bugs.

    Fixed a couple of bugs in Range::Object and Range::Extension,
    all with invalid input handling.

    Refactored Range::Object a bit. Changes include new parsing
    method, more abstraction; Range::Extension is now descended
    from Object instead of String.

    Test suite reworked too; it is more unified now and includes
    more thorough testing for different modules and input types.

0.01  Sat Aug 06 00:17:19 2011

    Started implementing Range::Object


t/tests.pl
sub run_tests {
    my $tests = shift;
    
    while ( @$tests ) {
        my $module = shift @$tests;
        next if @ARGV && !grep /$module/, @ARGV;
    
        my $subsort  = $module =~ /Serial/   ? sub { $a <=> $b }
                     :                         sub { $a cmp $b }
                     ;
    
        my @data     = @{ shift @$tests };
    
        # Execute custom code if it's there
        my $code = shift @data;
        $code->() if $code;
    
        # Range::Interval specific data
        my $interval;
        $interval = shift @data if $module =~ /Interval/;
    
        # Fetch invalid input data
        my @invalid_input = @{ shift @data };
    
        #
        # Invalid input test
        #
        SKIP: {
            skip "No invalid input tests for Range::String", 4
                if $module =~ /Range::String/;
    
            my $i = 0;
            for ( @invalid_input ) {
                $i++;
                my $range = $module =~ /Interval/
                          ? eval { $module->new($interval, $_) }
                          : eval { $module->new($_) }
                          ;
    
                fail "Didn't die on $module invalid input $i" unless $@;
                like $@, qr/Invalid input/, "$module invalid input $i";
            }
        }
    
        # Declare variables for consistency
        my ($range1,               $range2              );
        my (@range1_not_in_output, @empty               );
        my (@range1_list_output,   @range2_list_output  );
        my ($range1_string_output, $range2_string_output);
        my (@range1_short_output,  @range2_short_output );
        my ($range1_short_string,  $range2_short_string );
        my ($range1_stringify,     $range2_stringify    );
        my ($range1_size,          $range2_size         );
    
        for my $test ( qw( new add remove ) ) {
            # Fetch test data
            my @input         = @{ shift @data };
            my @in_list       = @{ shift @data };
            my @not_in_input  = sort $subsort @{ shift @data };
            my @not_in_output = sort $subsort @{ shift @data };
            my @range_output  = @{ shift @data };  # List   range() output
            my $range_output  =    shift @data  ;  # Scalar range() output
            my @short_output  = @{ shift @data };  # List   collapsed() output
            my $short_output  =    shift @data  ;  # Scalar collapsed() output
            my $size          =    shift @data  ;
    
            #
            # Test input
            #
            if ( $test eq 'new' ) {
                $range1 = $module =~ /Interval/
                        ? eval { $module->new($interval, @input) }
                        : eval { $module->new(@input) }
                        ;
    
                is     $@, '',          "$module $test eval $@";
                ok     defined $range1, "$module object created";
                isa_ok $range1,         $module;
            }
            else {
                eval { $range1->$test(@input) };
    
                is     $@, '', "$module $test eval $@";
            };
    
            #
            # scalar in() checks
            #
            ok $range1->in($_), "scalar $module->in($_) after $test()"
                for @in_list;
    
            #
            # list in() checks
            #
            @empty = $range1->in(@in_list);        # Should be empty
    
            is  scalar @empty, 0,  "list $module->in() after $test() no items";
            is_deeply \@empty, [], "list $module->in() after $test()    items";
    
            #
            # scalar not-in() checks
            #
            ok !$range1->in($_), "scalar not $module->in($_) after $test()"
                for @not_in_input;
    
            #
            # list not-in() checks
            #
            @range1_not_in_output = $range1->in(@not_in_input);
    
            is  scalar @range1_not_in_output, scalar @not_in_input,
                "list not $module->in() after $test() # items";
            is_deeply \@range1_not_in_output, \@not_in_input,
                "list not $module->in() after $test()   items";
    
            #
            # range() checks, list context
            #
            @range1_list_output = sort $subsort $range1->range();
            is_deeply \@range1_list_output, \@range_output,
                "list $module->range() after $test()";
    
            #
            # range() test, scalar context
            #
            $range1_string_output = $range1->range();
    
            is $range1_string_output, $range_output,
                "scalar $module->range() after $test()";
    
            #
            # collapse() test, list context
            #
            @range1_short_output = $range1->collapsed();
    
            is_deeply \@range1_short_output, \@short_output,
                "list $module->collapsed() after $test()";
    
            #
            # collapse() and stringify() tests, scalar context
            #
            $range1_short_string  = $range1->collapsed();
            $range1_stringify     = "$range1";
    
            is $range1_short_string, $short_output,
                "scalar $module->collapsed() after $test()";
            is $range1_short_string, $range1_stringify,
                "scalar $module->stringify() after $test()";
    
            #
            # size() tests
            #
            $range1_size = $range1->size();
            is $range1_size, $size, "$module->size() after $test()";
    
            #
            # trying to eat our own dog food
            #
            $range2 = $module =~ /Interval/
                    ? eval { $module->new($interval, $range1_string_output) }
                    : eval { $module->new($range1_string_output) }
                    ;
            is $@, '', "dog food $module->new() after $test() eval $@";
     
            @range2_list_output   = $range2->range();
            $range2_string_output = $range2->range();
            @range2_short_output  = $range2->collapsed();
            $range2_short_string  = $range2->collapsed();
            $range2_stringify     = "$range2";
            $range2_size          = $range2->size();
    
            is_deeply \@range2_list_output, \@range_output,
                "dog food $module->range() after $test() list   output";
            is $range2_string_output, $range_output,
                "dog food $module->range() after $test() string output";
            is_deeply \@range2_short_output, \@short_output,
                "dog food $module->collapsed() after $test() list output";
            is $range2_short_string, $short_output,
                "dog food $module->collapsed() after $test() string output";
            is $range2_stringify, $range2_short_string,
                "dog food $module->stringify_collapsed() after $test()";
            is $range2_size, $size,
                "dog food $module->size() after $test()";
    
            if ( $module eq 'Range::Interval' ) {
                my @military_range = $range1->military();
                my @expected_mil   = @{ shift @data };
    
                is_deeply \@military_range, \@expected_mil,
                    "$module military() after $test() list context";
    
                my $military_str   = $range1->military();
                my $expected_str   = shift @data;
    
                is $military_str, $expected_str,
                    "$module military() after $test() scalar context";
            };
        };
    };
}

1;


lib/Range/Serial.pm
package Range::Serial;

# This is basically what common::sense does, but without the pragma itself
# to remain compatible with Perls older than 5.8

use strict;
no  warnings;
use warnings qw(FATAL closed internal debugging pack malloc portable
                prototype inplace io pipe unpack deprecated glob digit
                printf reserved taint closure semicolon);
no  warnings qw(exec newline unopened);

use Carp;
use List::Util qw( first );

use base qw(Range::Object);

# Overload definitions

use overload q{""}    => 'stringify_collapsed',
             fallback => 1;

### PUBLIC INSTANCE METHOD ###
#
# Returns regex for matching unsigned integer numbers.
#

sub pattern {
    my ($self) = @_;

    my $delimiter = $self->delimiter();

    return qr/
                (?<! \A -)          # Can't have negative numbers
                \s*                 # optional greedy whitespace
                [0-9]+              # then one or more digits
                (?:                 # and finally this group consisting of
                    (?: $delimiter) # ... either delimiter
                    |               # or
                    (?: \s)         # whitespace
                    |               # or
                    \z              # end of string
                )                   # ... end of group
            /xms;
}

### PRIVATE INSTANCE METHOD ###
#
# Returns regex that is used to separate items in a range list.
# Default for Serial is comma (,) or semicolon (;).
#

sub separator {
    return qr/
                [,;]                # comma or semicolon
                \s*                 # if there is whitespace, eat it all
            /xms;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns range delimiter; Serial uses default dash (-).
#

sub delimiter  { '-' }

############## PRIVATE METHODS BELOW ##############

### PRIVATE INSTANCE METHOD ###
#
# Tests if a sigle value is in current range. Serial uses numeric
# comparison.
#

sub _search_range {
    my ($self, $value) = @_;

    return
        first {
                ref($_) ? (($value >= $_->{start}) && ($value <= $_->{end}))
                :         $_ == $value
              }
              @{ $self->{range} };
}

### PRIVATE INSTANCE METHOD ###
#
# Returns sorted list of all single items within current range.
# Serial uses numeric sorting.
#
# Works in list context only, croaks if called otherwise.
#

sub _sort_range {
    my ($self, @range) = @_;

    croak "_sort_range can only be used in list context"
        unless wantarray;

    return sort { $a <=> $b } @range ? @range : $self->_full_range();
}

### PRIVATE INSTANCE METHOD ###
#
# Tests if two values are equal.
#

sub _equal_value {
    my ($self, $first, $last) = @_;

    return !!($first == $last);
}

### PRIVATE INSTANCE METHOD ###
#
# Tests if two values are consequent.
#

sub _next_in_range {
    my ($self, $first, $last) = @_;

    return !!($last == $first + 1);
};

1;

__END__

=pod

=head1 NAME

Range::Serial - Implements ranges of integer identificators

=head1 SYNOPSIS

 use Range::Serial;
 
 # Create a new range
 my $range = Range::Serial->new('1-10, 12, 15', 17..20);
 
 # Test if a value is in range
 print "in range\n"     if  $range->in(5);
 print "not in range\n" if !$range->in(11);
 
 # Add values to range
 $range->add(11, '13-14;16');
 
 # Get full list of values
 my @list = $range->range();
 print join q{ }, @list;         # Prints 1 2 3 4 5 ... 20
 
 # Get collapsed string representation
 my $string = $range->range();
 print "$string\n";              # Prints '1-20'
 
 # Get range size
 my $size = $range->size();
 print "$size";                  # Prints: 20
 
=head1 DESCRIPTION

This module implements ranges of positive integer numbers that can be used
as object identifiers.

=head1 METHODS

See L<Range::Object>.

=head1 BUGS AND LIMITATIONS

Only positive integer numbers are supported. No backward ranges are
supported, i.e. while '1-10' is valid range, '10-1' is not.

There are no known bugs in this module. Please report problems to author,
patches are welcome.

=head1 AUTHOR

Alexander Tokarev E<lt>tokarev@cpan.orgE<gt>.

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011 Alexander Tokarev.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.


lib/Range/Object.pm
package Range::Object;

# This is basically what common::sense does, but without the pragma itself
# to remain compatible with Perls older than 5.8

use strict;

no  warnings;
use warnings qw(FATAL closed internal debugging pack malloc portable
                prototype inplace io pipe unpack deprecated glob digit
                printf reserved taint closure semicolon);
no  warnings qw(exec newline unopened);

use Carp;
use List::Util qw( first );

### PACKAGE VARIABLE ###
#
# Version of this module.
#

# This is for compatibility with older Perls
use vars qw( $VERSION );

$VERSION = '0.92';

### PUBLIC CLASS METHOD (CONSTRUCTOR) ###
#
# Initializes new instance of $class from @input_range.
#

sub new {
    my ($class, @input_range) = @_;

    my $self = bless { range => [] }, $class;

    return $self->add(@input_range);
}

### PUBLIC INSTANCE METHOD ###
#
# Validates @input_range of items and adds them to internal storage.
#

sub add {
    my ($self, @input_range) = @_;

    # Nothing to do
    return $self unless @input_range;

    my @validated_input = $self->_validate_and_expand(@input_range);

    # Expand existing range and overlay the new one
    my %existing_values = map {; "$_" => 1 } $self->_full_range();
    @existing_values{ @validated_input } = (1) x @validated_input;

    # Collapse resulting hash and replace current range with new values
    $self->{range} = [ $self->_collapse_range( keys %existing_values ) ];

    return $self;
}

### PUBLIC INSTANCE METHOD ###
#
# Removes items in @input_range from internal storage.
#

sub remove {
    my ($self, @input_range) = @_;

    # Nothing to do
    return $self unless @input_range;

    my @validated_input = $self->_validate_and_expand(@input_range);

    # Expand existing range and remove what needs to be removed
    my %existing_values = map {; "$_" => 1 } $self->_full_range();
    delete @existing_values{ @validated_input };

    # Collapse resulting hash and replace current range with new values
    $self->{range} = [ $self->_collapse_range( keys %existing_values ) ];

    return $self;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns sorted array or string representation of internal storage.
# In scalar context it can use optional list separator instead of
# default one.
#

sub range {
    my ($self, $separator) = @_;

    return wantarray    ? $self->_sort_range()
         :                $self->stringify($separator)
         ;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns sorted and collapsed representation of internal storage.
# In list context, resulting list consists of separate values and/or
# range hashrefs with three elements: start, end and count.
# In scalar context, result is a string of separate values and/or
# ranges separated by value returned by delimiter() method.
# Optional list separator can be used instead of default one in
# scalar context.
#

sub collapsed {
    my ($self, $separator) = @_;

    return wantarray    ? @{ $self->{range} }
         :                $self->stringify_collapsed($separator)
         ;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns the number of separate items in internal storage.
#

sub size {
    my ($self) = @_;

    my $size = 0;
    for my $item ( @{ $self->{range} } ) {
        $size += ref $item ? $item->{count} : 1;
    };

    return $size;
}

### PUBLIC INSTANCE METHOD ###
#
# Tests if items of @input_range are matching items in our internal storage.
# Returns true/false in scalar context, list of mismatching items in list
# context.
#

sub in {
    my ($self, @input_range) = @_;

    my @validated_range = $self->_validate_and_expand(@input_range);

    if ( wantarray ) {
        # This should be reasonably fast
        return grep { !defined $self->_search_range("$_") } @validated_range;
    }
    else {
        # This should be even faster
        return defined first {
                                 my $result = $self->_search_range("$_");
                                   !defined $result ? ''
                                 : $result == 0     ? 1
                                 :                    $result
                             }
                             @validated_range;
    };

    return;     # Just for fun
}

### PUBLIC INSTANCE METHOD ###
#
# Returns string representation of all items in internal storage (sorted).
#

sub stringify {
    my ($self, $separator) = @_;

    $separator ||= $self->_list_separator();

    return join $separator, $self->_sort_range();
}

### PUBLIC INSTANCE METHOD ###
#
# Returns string representation of collapsed current range.
#

sub stringify_collapsed {
    my ($self, $separator) = @_;

    $separator ||= $self->_list_separator();

    my @collapsed_range
        = map {
                ref($_) ? $self->_stringify_range( $_->{start}, $_->{end} )
                :         "$_"
              }
              @{ $self->{range} };

    return join $separator, @collapsed_range;
}

### PUBLIC INSTANCE METHOD ###
#
# Returns regex that is used to validate range items. Regex should
# include patterns both for separate disjointed items and contiguous
# ranges.
# Default pattern matches everything.
#

sub pattern { qr/.*?/xms }

### PUBLIC INSTANCE METHOD ###
#
# Returns regex that is used to separate items in a range list.
# Default is no separator.
#

sub separator { qr//xms }

### PUBLIC INSTANCE METHOD ###
#
# Returns default range delimiter for current class.
#

sub delimiter { '-' }

############## PRIVATE METHODS BELOW ##############

### PRIVATE INSTANCE METHOD ###
#
# Returns default list separator for use with stringify() and
# stringify_collapsed()
#

sub _list_separator { q{,} }

### PRIVATE INSTANCE METHOD ###
#
# Validates and unpacks input @input_range of items, returns full list.
#

sub _validate_and_expand {
    my ($self, @input_range) = @_;

    # Nothing to do
    return unless @input_range;

    # We need the patterns
    my $pattern    = $self->pattern();
    my $separator  = $self->separator();

    # We use hash to avoid duplicates
    my %temp = ();

    # Expand and validate items in @input_range; add them if all is OK
    ITEM:
    while ( @input_range ) {
        my $item = shift @input_range;

        if ( $separator && $item =~ $separator ) {
            unshift @input_range, split $separator, $item;
            next ITEM;
        };

        croak "Invalid input: $item"
            if !defined $item || $item eq '' || $item !~ /$pattern/;

        # Default expansion mechanism is Perl range operator (..)
        my @items = eval { $self->_explode_range($item) };
        croak "Invalid input item '$item': $@" if $@;

        # Store result to temp hash, avoiding duplicates
        @temp{ @items } = (1) x @items;
    };

    # We need to sort items because order matters
    my @result = $self->_sort_range( keys %temp );

    return @result;
}

### PRIVATE INSTANCE METHOD ###
#
# Explodes stringified range of items using Perl range operator.
#

sub _explode_range {
    my ($self, $string) = @_;

    my $delimiter = $self->delimiter();

    # Shortcut
    for ($string) {
        # Remove whitespace and normalize separators
        s/\s+//g;
        s/;/,/g;

        # Replace delimiters with (..) honoring qw() constructs
               s{  \)  \s* $delimiter \s* qw\( }     {)..qw(}gx
        unless s{ (\d) \s* $delimiter \s* (\d) }     {$1..$2}gx;
    };

    my $items_ref = eval "[$string]";

    return @$items_ref;
}

### PRIVATE INSTANCE METHOD ###
#
# Tests if a sigle value is in current range.
#

sub _search_range {
    my ($self, $value) = @_;

    return first {
                    ref($_) ? $self->_is_in_range_hashref($_, $value)
                    :         $self->_equal_value("$_", $value)
                 }
                 @{ $self->{range} };
}

### PRIVATE INSTANCE METHOD ###
#
# Tests if a single value is within boundaries of collapsed range item.
# Default method uses string comparison.
#

sub _is_in_range_hashref {
    my ($self, $range_ref, $value) = @_;

    return (    ($value ge $range_ref->{start})
             && ($value le $range_ref->{end})
           );
}

### PRIVATE INSTANCE METHOD ###
#
# Returns sorted list of all single items within current range.
# Default sort is string-based.
#
# Works in list context only, croaks if called otherwise.
#

sub _sort_range {
    my ($self, @range) = @_;

    croak "Internal error: _sort_range can only be used in list context"
        unless wantarray;

    return sort { $a cmp $b } @range ? @range : $self->_full_range();
}

### PRIVATE INSTANCE METHOD ###
#
# Returns full list of items in current range.
#

sub _full_range {
    my ($self) = @_;

    croak "Internal error: _full_range can only be used in list context"
        unless wantarray;

    my $delimiter = $self->delimiter();

    return map {
                   ref($_) ? $self->_explode_range( $_->{start}
                                                    . $delimiter
                                                    . $_->{end}
                                                  )
                   :         "$_"
               }
               @{ $self->{range} };
}

### PRIVATE INSTANCE METHOD ###
#
# Returns collapsed list of current range items. Separate items are
# returned as is, contiguous ranges are collapsed and returned as
# hashrefs { start => $start, end => $end, count => $count }.
#
# Works in list context only, croaks if called otherwise.
#

sub _collapse_range {
    my ($self, @range) = @_;

    croak "Internal error: _collapse_range can only be used in list context"
        unless wantarray;

    my ($first, $last, $count, @result);

    ITEM:
    for my $item ( $self->_sort_range(@range) ) {
        # If $first is defined, it means range has started
        if ( !defined $first ) {
            $first = $last = $item;
            $count = 1;
            next ITEM;
        };

        # If $last immediately preceeds $item in range,
        # $item becomes next $last
        if ( $self->_next_in_range($last, $item) ) {
            $last = $item;
            $count++;
            next ITEM;
        };

        # If $item doesn't follow $last and $last is defined,
        # it means current contiguous range is complete
        if ( !$self->_equal_value($first, $last) ) {
            push @result, {
                              start => $first,
                              end   => $last,
                              count => $count,
                          };
            $first = $last = $item;
            $count = 1;
            next ITEM;
        };

        # If $last wasn't defined, range was never contiguous
        push @result, "$first";
        $first = $last = $item;
        $count = 1;
    }

    # We're here when last item has been processed
    push @result,
        $self->_equal_value($first, $last)    ? "$first"
        :                                       {
                                                    start => $first,
                                                    end   => $last,
                                                    count => $count,
                                                }
        ;

    return @result;
}

### PRIVATE INSTANCE METHOD ###
#
# Tests if two values are equal. This method has to be overridden.
#

sub _equal_value {
    my ($self, $first, $second) = @_;

    croak "Internal error: Can't use _equal_value with Range::Object";
}

### PRIVATE INSTANCE METHOD ###
#
# Tests if two values are consequent. This method has to be overridden.
#

sub _next_in_range {
    my ($self, $first, $second) = @_;

    croak "Internal error: Can't use _next_in_range with Range::Object";
}

### PRIVATE INSTANCE METHOD ###
#
# Returns stringified representation of a range within $first and $last
# boundaries.
#

sub _stringify_range {
    my ($self, $first, $last) = @_;

    my $delimiter = $self->delimiter();

    return $first . $delimiter . $last;
}

1;

__END__

=pod

=head1 NAME

Range::Object - Basic facilities for manipulating different kinds of object ranges

=head1 SYNOPSIS

This module is not to be used directly. See L<Range::Serial>,
L<Range::Strings>, L<Range::Extension>, L<Range::DigitString>,
L<Range::Date> and L<Range::Interval>.

=head1 DESCRIPTION

This module provides abstract methods for Range::* family of modules.

The purpose of these modules is to provide unified interface for storing,
retrieving and testing for existence of different kinds of objects and
object ranges.

In terms of this namespace, a range is defined as a set of items, either
disjointed (individual) or contiguous, or a combination of separate items
and ranges. Intersecting or adjacent ranges are not supported directly and
will be collapsed silently into wider contiguous range.

Although Range::Object descendant module can store any number of separate
values (objects) and ranges, it is optimized for storing contiguous ranges
of arbitrary length with minimal memory and storage footprint; the other
effect of this being fast serialization and deserialization of Range::Object
instances. It cannot come without cost though; Range::Object uses more
CPU cycles than similar hash-based modules.

Good application for this kind of object storage can be an implementation
of user permission tables for large number of objects, especially if such
permissions are typically assigned in large contiguous ranges. For
example, if User has read permission for objects 1-10000 and write
permission for objects 1-100, 200-300 and 1000-9999, storing these
identificators as hash keys is memory expensive, and can become
prohibive when number of tables and users increase. Compared to that
approach, Range::Object can be a reasonable compromise between memory
and CPU utilization.

=head1 METHODS

=over 4

=item new(@list)

Takes a @list of arguments, expands them and creates range object.
The following separators are allowed: comma (,), semicolon (;) and
dash (-) which means literal range between two items.

=item add(@list)

Adds items to range using the same rules as with new(). In fact, new()
will call add() after initialization.

=item remove(@list)

Removes items in @list from current range. As with add(), @list can
contain individual items as well as stringified item ranges.

=item size()

Returns the number of single individual items in the range.

=item in(@list)

Checks if items in @list are within our current range. In scalar context,
returns true or false; in list context returns items from expanded @list
that are not in range or empty list if they are all in range.

N.B: it means that results in scalar and list context are opposite: true
in scalar context would be empty list in list context which evaluates to
false.

=item range( [$separator] )

Returns the list of items in current range in list context, or
range string in scalar context. Stringified version of range() is
join()ed using optional $separator or default comma (,).

=item collapsed( [$separator] )

Returns collapsed list of items in current range in list context, or
stringified version of collapsed list in scalar context.

Collapsed list consists of either individual items or hashrefs
representing ranges in the following format:
    {
        start => <starting item>,
        end   => <ending item>,
        count => <number of items>
    }

In stringified version, ranges are delimited with whatever character or
string is returned by delimiter(). Optional $separator can specify a
character or string to be used as list separator for range items.

=item stringify( [$separator] )

Returns string representation of current range. This method can be
used instead of range() in scalar context to provide the same results.

Optional $separator can specify a character or string to be used
as list separator for range items.

=item stringify_collapsed( [$separator] )

Returns string representation of collapsed current range, following the
same rules as used by new() for expanding ranges. In fact, the output of
stringify_collapsed() can be fed back to new() to create an exact copy
of the current range.  This method also can be used instead of collapsed()
in scalar context to provide the same results.

Optional $separator specifies a character or string to be used as list
separator for range items.

=item pattern()

Returns regex that is used to validate range items.

=item separator()

Returns regex that is used to split input range items.

=item delimiter()

Returns range delimiter for current class, default is dash (-).

=back

=head1 DIAGNOSTICS

=over 4

=item Invalid input: 'foo'

add() will fail with this message if input list item or range is not
matched by pattern defined by current module.

=item Invalid input item: 'bar'

add() will fail with this message if input list *range* item is not
matched by pattern defined by current module.

=back

=head1 DEPENDENCIES

This module is dependent on the following standard modules:
L<Carp>, L<List::Util>.

=head1 BUGS AND LIMITATIONS

Addition/removal operations are quite resource heavy in present implementation
as they require complete unpacking/collapsing of the current range for
each add or remove. There are no plans for optimization on this part, since
permission tables are usually very long lived and relatively rarely changed.
Patches and ideas are welcome though.

Only forward ranges are supported, i.e. starting value should be less than
or equal to ending value.

There are no known bugs in this module. Please report problems to author,
patches are welcome.

=head1 AUTHOR

Alexander Tokarev E<lt>tokarev@cpan.orgE<gt>.

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011 by Alexander Tokarev.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.
